### PR TITLE
tpm2_rsaencrypt|decrypt #1159

### PR DESCRIPTION
**Pull request by GlovePuppet**
Extend tpm2_rsaencrypt & tpm2_rsadecrypt to support setting the padding scheme

### CODE REVIEW

File: tools/tpm2_rsadecrypt.c [hunk 1]
```diff
@@ -123,6 +148,9 @@ static bool on_option(char key, char *value) {
         ctx.flags.o = 1;
         break;
     }
+    case 's':
+        ctx.scheme.scheme = str_to_pad(value);
```
Comment (williamcroberts) [inline tools/tpm2_rsadecrypt.c]: tpm2_alg_util.c: tpm2_alg_util_from_optarg(value, tpm2_alg_util_flags_enc_scheme) should do this and all error checking.

File: tools/tpm2_rsaencrypt.c [hunk 2]
```diff
@@ -132,6 +162,11 @@ static bool init(ESYS_CONTEXT *context) {
         return false;
     }
 
+    if(ctx.scheme.scheme == TPM2_ALG_ERROR) {
```
Comment (williamcroberts) [inline tools/tpm2_rsaencrypt.c]: do the error checking of this in the optarg handler, no need to defer it.

File: tools/tpm2_rsadecrypt.c (hunk 1 above)
Comment (GlovePuppet) [reply tools/tpm2_rsadecrypt.c]: I looked at using that but I wasnt overly happy
-We would need tpm2_alg_util_flags_misc | tpm2_alg_util_flags_enc_scheme
-We have to filter out ecdh (it's in the tpm2_alg_util_flags_enc_scheme group but not a padding scheme option)
If that's still ok, I will make the changes

Comment (williamcroberts) [reply tools/tpm2_rsadecrypt.c]: apply this and work it up:
```diff
diff --git a/lib/tpm2_alg_util.c b/lib/tpm2_alg_util.c
index a3f15193cca8..4ec7e5e370a1 100644
--- a/lib/tpm2_alg_util.c
+++ b/lib/tpm2_alg_util.c
@@ -97,8 +97,8 @@ static void tpm2_alg_util_for_each_alg(alg_iter iterator, void *userdata) {
         { .name = "ecschnorr", .id = TPM2_ALG_ECSCHNORR, .flags = tpm2_alg_util_flags_sig },
 
         // Assyemtric Encryption Scheme
-        { .name = "oaep", .id = TPM2_ALG_OAEP, .flags = tpm2_alg_util_flags_enc_scheme },
-        { .name = "rsaes", .id = TPM2_ALG_RSAES, .flags = tpm2_alg_util_flags_enc_scheme },
+        { .name = "oaep", .id = TPM2_ALG_OAEP, .flags = tpm2_alg_util_flags_enc_scheme | tpm2_alg_util_flags_rsa_scheme},
+        { .name = "rsaes", .id = TPM2_ALG_RSAES, .flags = tpm2_alg_util_flags_enc_scheme | tpm2_alg_util_flags_rsa_scheme },
         { .name = "ecdh", .id = TPM2_ALG_ECDH, .flags = tpm2_alg_util_flags_enc_scheme },
 
 
@@ -123,7 +123,7 @@ static void tpm2_alg_util_for_each_alg(alg_iter iterator, void *userdata) {
         { .name = "keyedhash", .id = TPM2_ALG_KEYEDHASH, .flags = tpm2_alg_util_flags_base },
 
         // Misc
-        { .name = "null", .id = TPM2_ALG_NULL, .flags = tpm2_alg_util_flags_misc },
+        { .name = "null", .id = TPM2_ALG_NULL, .flags = tpm2_alg_util_flags_misc | tpm2_alg_util_flags_rsa_scheme},
     };
 
     size_t i;
diff --git a/lib/tpm2_alg_util.h b/lib/tpm2_alg_util.h
index e3783dfccab6..2caa72e22e3e 100644
--- a/lib/tpm2_alg_util.h
+++ b/lib/tpm2_alg_util.h
@@ -49,6 +49,7 @@ enum tpm2_alg_util_flags {
     tpm2_alg_util_flags_base        = 1 << 8,
     tpm2_alg_util_flags_misc        = 1 << 9,
     tpm2_alg_util_flags_enc_scheme  = 1 << 10,
+    tpm2_alg_util_flags_rsa_scheme  = 1 << 11,
     tpm2_alg_util_flags_any         = ~0
 };
```

Comment (GlovePuppet) [reply tools/tpm2_rsadecrypt.c]: TY

File: tools/tpm2_rsadecrypt.c [hunk 3]
```diff
@@ -60,19 +61,19 @@ struct tpm_rsadecrypt_ctx {
     char *key_auth_str;
     const char *context_arg;
     tpm2_loaded_object key_context_object;
+    TPMT_RSA_DECRYPT scheme;
 };
 
 tpm_rsadecrypt_ctx ctx = {
-    .auth = { .session_data = TPMS_AUTH_COMMAND_INIT(TPM2_RS_PW) }
+    .auth = { .session_data = TPMS_AUTH_COMMAND_INIT(TPM2_RS_PW) },
+    .scheme = {.scheme = TPM2_ALG_RSAES}
```
Comment (williamcroberts) [inline tools/tpm2_rsadecrypt.c]: spaces between {}

File: tools/tpm2_rsaencrypt.c [hunk 4]
```diff
@@ -49,27 +50,27 @@ struct tpm_rsaencrypt_ctx {
     TPM2B_PUBLIC_KEY_RSA message;
     char *output_path;
     char *input_path;
+    TPMT_RSA_DECRYPT scheme;
 };
 
 static tpm_rsaencrypt_ctx ctx = {
-    .context_arg = NULL
+    .context_arg = NULL,
+    .scheme = {.scheme = TPM2_ALG_RSAES}
```
Comment (williamcroberts) [inline tools/tpm2_rsaencrypt.c]: spaces between {}

File: lib/tpm2_alg_util.c [hunk 5]
```diff
@@ -97,8 +97,8 @@ static void tpm2_alg_util_for_each_alg(alg_iter iterator, void *userdata) {
         { .name = "ecschnorr", .id = TPM2_ALG_ECSCHNORR, .flags = tpm2_alg_util_flags_sig },
 
         // Assyemtric Encryption Scheme
-        { .name = "oaep", .id = TPM2_ALG_OAEP, .flags = tpm2_alg_util_flags_enc_scheme },
-        { .name = "rsaes", .id = TPM2_ALG_RSAES, .flags = tpm2_alg_util_flags_enc_scheme },
+        { .name = "oaep", .id = TPM2_ALG_OAEP, .flags = tpm2_alg_util_flags_enc_scheme | tpm2_alg_util_flags_rsa_scheme},
+        { .name = "rsaes", .id = TPM2_ALG_RSAES, .flags = tpm2_alg_util_flags_enc_scheme | tpm2_alg_util_flags_rsa_scheme },
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: spaces between } here and throughout.

File: man/tpm2_rsadecrypt.1.md [hunk 6]
```diff
@@ -42,7 +42,7 @@ The key referenced by key-context is **required** to be:
 
     Output file path, record the decrypted data.
 
-  * **-s**, **--pad-scheme**=_SCHEME_:
+  * **-g**, **--halg**=_ALGORITHM_:
```
Comment (williamcroberts) [inline man/tpm2_rsadecrypt.1.md]: it's not a hashing algorithm aka "halg". So something like --algorithm or --scheme would be better here.

File: man/tpm2_rsaencrypt.1.md [hunk 7]
```diff
@@ -35,7 +35,7 @@ The key referenced by key-context is **required** to be:
     xxd compatible hexdump to stdout. If a file is specified, raw binary
     output is performed.
 
-  * **-s**, **--pad-scheme**=_SCHEME_:
+  * **-g**, **--halg**=_ALGORITHM_:
```
Comment (williamcroberts) [inline man/tpm2_rsaencrypt.1.md]: same

**Comment by GlovePuppet** [comment]
Hopefully I have fixed the git issue & the latest review comments

**Pull request merged**